aiomqtt/client.py

```diff
@@ -1052,21 +1102,30 @@ async def __aexit__(
         exc: BaseException | None,
         tb: TracebackType | None,
     ) -> None:
-        """Try to gracefully disconnect from the broker."""
-        try:
-            if self._early_out_on_disconnected():
-                return
-            try:
-                await self.disconnect()
-            except MqttError as error:
-                # We tried to be graceful. Now there is no mercy.
-                self._logger.warning(
-                    f'Could not gracefully disconnect due to "{error}". Forcing'
-                    " disconnection."
-                )
-        finally:
-            self._force_disconnect()
+        """Disconnect from the broker."""
+        if self._disconnected.done():
+            # Return early if the client is already disconnected
             self._lock.release()
+            if (exc := self._disconnected.exception()) is not None:
+                # If the disconnect wasn't intentional, raise the error that caused it
+                raise exc
+            return
+        # Try to gracefully disconnect from the broker
+        rc = self._client.disconnect()
+        if rc == mqtt.MQTT_ERR_SUCCESS:
+            # Wait for acknowledgement
+            await self._wait_for(self._disconnected, timeout=None)
```

Comment:
If it times out here, the lock will not be released, so my suggestion is to change the release lock to be executed anyway in aexit.

Reply:
Thanks for checking the changes! 😊 If we release the lock, we can then call `__aenter__` while the client is (potentially) still connected. I'm not sure what's best here.

In any case, I aimed for having equal logic to before in this PR. I hope that we can simplify the internal state when we remove `connect` and `disconnect`, so that such things are clearer. Right now it's slightly confusing having `_connected`, `_disconnected`, _and_ `_lock` and what these three should specify.